### PR TITLE
v8.0.0: hover/drag_drop/wait_for_stable + 27 new adapters + sheets-mode foundation

### PR DESCRIPTION
**Pull request by esokullu**
## Summary

Two-phase change comparing WebBrain's extension architecture against FSB and acting on the gaps.

**Phase 1 (commit `3c3f57d`) — v8.0.0:** Three new tools and 27 new site adapters.

**Phase 2.1 (commit `37af410`) — sheets-mode foundation:** Pure-helpers + plumbing for `read_sheet` / `fill_sheet` coming in v8.1.0. No LLM-facing tools yet.

## Phase 1: new tools

| Tool | Chrome | Firefox | Rationale |
|---|---|---|---|
| **`hover`** | CDP-trusted `Input.dispatchMouseEvent` mouseMoved | Synthetic `mouseenter`/`pointerover` (isTrusted=false) | Reveal-on-hover menus (GitHub three-dot, Linear card actions) that the AX tree doesn't see until cursor enters |
| **`drag_drop`** | CDP pointer-event sequence with N waypoints | Synthetic pointer + HTML5 dnd with constructed DataTransfer | Trello/Linear/Notion card reorder, crop handles, sliders — no way to do this through clicks |
| **`wait_for_stable`** | MutationObserver + fetch/XHR-idle polling | Same | Closes the gap between `wait_for_element` ("did X appear") and what the agent actually needs after `navigate` / `set_field({submit:true})` ("is the page done shuffling") |

**`right_click` was considered and explicitly NOT added:** CDP fires the `contextmenu` DOM event but cannot trigger the OS-native menu, and sites that have a JS `oncontextmenu` handler almost all expose the same actions through a discoverable "..." button. Adding the tool would just give the model another way to misfire.

## Phase 1: site adapters (30 → 57)

| Category | New sites |
|---|---|
| Travel (7) | airbnb, booking, expedia, google-maps, google-flights, kayak, opentable |
| E-commerce (4) | ebay, walmart, target, etsy |
| Productivity (3) | outlook, google-sheets, trello |
| Social (3) | instagram, tiktok, facebook |
| Finance specifics (3) | coinbase, robinhood, tradingview (on top of existing `finance-generic` regex catch-all) |
| Coding practice (2) | leetcode, hackerrank |
| Job portals (2) | greenhouse, workday |
| Messaging (3) | discord, whatsapp-web, telegram |

All adapters written WebBrain-style: ~15-25 lines, observed-failure-driven, no encoded selectors. Only one adapter fires per URL match so prompt cost is fixed regardless of total count.

## Phase 2.1: sheets-mode foundation

Google Sheets and Excel Online render the cell grid to `<canvas>`. There's no DOM for cells, so `get_accessibility_tree` is empty for the grid and `read_page` only surfaces the active cell's value via the formula bar. Bulk read/write needs to go around the canvas.

**Approach picked: keyboard + clipboard.** Uses the user's existing browser session (no OAuth flow to add), works identically on both Sheets and Excel Online with one code path.

**This commit:** scaffolding only — pure helpers tested, manifest permissions added, backend stubs in place. **No tools exposed to the LLM yet.** Phase 2.2 implements Google Sheets read; 2.3 write; 2.4 Excel Online; 2.5 exposes the tools and ships v8.1.0.

- `parseA1` — A1-notation → `{row, col, rowCount, colCount, sheet?}`. Handles ranges, whole-column (`A:A`), whole-row (`1:1`), and `Sheet2!A1` / `'My Sheet'!A1` prefixes. Rejects ambiguous single-letter input like `"A"` with a clear error — silently parsing that as `A1` would be the worst failure mode.
- `colLettersToIndex` / `indexToColLetters` — base-26 column math, round-tripped over 0–1000.
- `valuesToTsv` / `tsvToValues` — RFC-4180-style TSV with quoted cells for tab/newline/quote, matching what Sheets and Excel use for clipboard copy/paste.
- `detectSheetSite` — URL → `"google-sheets"` | `"excel-online"` | `null`.
- Clipboard snapshot/restore helpers backing the future read/write paths.

**Apple Numbers iCloud:** explicitly deferred to v8.2.0+. Better to ship Sheets + Excel solid than three half-cooked.

## Test coverage

- **128 tests** before this PR
- **179 tests** after (51 new): A1 parsing, column-name math (incl. 0–1000 roundtrip), TSV (de)serialization edge cases (quoted cells, embedded tabs/newlines/quotes, CRLF, empty rows), site detection, chrome/firefox parity
- All pass

## Files touched

**Phase 1** (17 files):
- `src/chrome/src/agent/{tools,agent,adapters}.js`
- `src/chrome/src/content/content.js`
- `src/chrome/ARCHITECTURE.md`
- `src/firefox/src/agent/{tools,agent,adapters}.js`
- `src/firefox/src/content/content.js`
- `src/firefox/ARCHITECTURE.md`
- 6 version-bump files (manifests × 3, package.json × 2, settings.js × 2)

**Phase 2.1** (6 files):
- `src/chrome/src/agent/sheets-tools.js` (NEW)
- `src/firefox/src/agent/sheets-tools.js` (NEW)
- All three manifests (clipboardRead + clipboardWrite added)
- `test/run.js` (sheets-tools test suite)

## Test plan

- [x] `npm test` — 179/179 passing
- [x] Sanity: tools load cleanly under Node (Chrome + Firefox copies)
- [x] Sanity: adapters parse to 57 entries on both browsers
- [ ] Manual smoke: load unpacked Chrome build, hover over a GitHub three-dot menu, verify the reveal happens
- [ ] Manual smoke: load unpacked Chrome build, drag a Trello card between lists, verify it moves
- [ ] Manual smoke: navigate, call `wait_for_stable`, verify it returns `stable:true`
- [ ] Manual smoke: visit each new adapter URL pattern, verify `getActiveAdapter(url)` returns the right adapter

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/chrome/src/agent/agent.js [hunk 1]
```diff
@@ -4764,6 +4764,134 @@ Rules: no prose intro, no conclusion, no "this screenshot shows...", no layout d
...
+        // If the destination scroll moved the viewport, the source coords are
+        // now stale — re-resolve source one more time. This is the failure
+        // mode that broke long-list Trello drags on the first iteration.
+        const fromResp2 = await chrome.tabs.sendMessage(tabId, {
+          target: 'content', action: 'ax_resolve_rect', params: { ref_id: fromRef },
+        });
+        const from = (fromResp2 && fromResp2.success) ? fromResp2 : fromResp;
+
+        await cdpClient.attach(tabId);
+        const x1 = from.x, y1 = from.y, x2 = toResp.x, y2 = toResp.y;
```
[the remark this answers is not in the document]
Comment (esokullu) [reply src/chrome/src/agent/agent.js]: Confirmed and fixed in baad53b.

Replaced the three-call `ax_resolve_rect` dance with a single new content-script handler `ax_resolve_two_rects` that:
1. Scrolls source into view
2. Scrolls dest into view
3. Measures BOTH rects in the same post-scroll viewport snapshot

`drag_drop` now uses the single round-trip. The previous code's re-resolve scrolled source back into view and invalidated the dest coords we'd just saved — exactly the failure mode you described.

Added a defensive check on `inViewport` for both ends. If source and dest are far enough apart that the final viewport (centered on dest) puts source off-screen, the drag is refused with a clear error rather than going through the motions of CDP events at off-screen coordinates that don't land. That's a pre-existing limitation surfaced honestly.

**Comment by esokullu** [comment]
Both Codex findings addressed in baad53b:

**P2 — adapter ordering.** `finance-generic` moved to AFTER the specific finance adapters (stripe / coinbase / robinhood / tradingview) in both chrome and firefox. Verified with runtime checks: `coinbase.com` → `coinbase`, `robinhood.com` → `robinhood`, `tradingview.com` → `tradingview`, `chase.com` → `finance-generic`. Also rewrote the now-misleading "Match before finance-generic" comment that claimed correct ordering it didn't have.

**P1 — drag_drop scroll race.** Replaced the three-call resolve-dance with a single new content-script handler `ax_resolve_two_rects` that scrolls both elements and measures both rects in the same post-scroll viewport snapshot. Plus a defensive inViewport check that surfaces a clear error if source and dest are too far apart to fit on screen at the same time.

179 tests still pass.

File: src/chrome/src/agent/agent.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (esokullu) [reply src/chrome/src/agent/agent.js]: @codex pls check again

**Comment by esokullu** [comment]
@codex pls check again

File: src/chrome/src/content/content.js [hunk 2]
```diff
@@ -1678,6 +1678,213 @@
...
+          if (checkNetwork && !window.__wbNetIdleInstalled) {
+            try {
+              window.__wbNetIdleInstalled = true;
+              window.__wbInFlight = 0;
+              const origFetch = window.fetch;
+              if (typeof origFetch === 'function') {
+                window.fetch = function patchedFetch() {
+                  window.__wbInFlight = (window.__wbInFlight | 0) + 1;
+                  return origFetch.apply(this, arguments).finally(() => {
+                    window.__wbInFlight = Math.max(0, (window.__wbInFlight | 0) - 1);
```
[the remark this answers is not in the document]
Comment (esokullu) [reply src/chrome/src/content/content.js]: Confirmed and fixed in 2cdf038.

You're right: patching `window.fetch` / XHR in the isolated content script observed nothing because the page's calls happen in MAIN world with its own copies of those globals.

Fixed by injecting a `<script>` element with the patches inline — that runs in the page's MAIN world and observes real traffic. The counter publishes its value to `document.documentElement.dataset.__wbInflight`; the content script reads the dataset attribute (DOM is shared across worlds).

Strict-CSP pages refuse the inline `<script>`. The handler detects that (dataset never gets set), returns `networkObserved: false` on the response, bypasses the netIdle gate, and decides stability on MutationObserver alone. The timeout hint explicitly says network couldn't be observed so the agent / user knows the stability claim is weaker.

Same fix applied to firefox content.js (its content scripts have the same isolated-compartment behavior).

File: src/chrome/src/agent/sheets-tools.js [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+  // failure mode — it reads/writes the wrong cell with no error. The
+  // explicit syntax for whole column is "A:A".
+  if (!hasEnd && (!sCol || !sRow)) {
+    throw new Error(`parseA1: incomplete start address in ${JSON.stringify(range)} — for a single cell pass both column and row (e.g. "A1"); for a whole column use "A:A"; for a whole row use "1:1".`);
+  }
+
+  const startCol = sCol ? colLettersToIndex(sCol) : 0;
+  const startRow = sRow ? parseInt(sRow, 10) - 1 : 0;
+  const endCol = hasEnd && eCol ? colLettersToIndex(eCol) : (hasEnd ? null : startCol);
+  const endRow = hasEnd && eRow ? parseInt(eRow, 10) - 1 : (hasEnd ? null : startRow);
```
[the remark this answers is not in the document]
Comment (esokullu) [reply src/chrome/src/agent/sheets-tools.js]: Fixed in 2cdf038.

Added an explicit `_validateRowStr` helper that's called on both the start row and the end row before zero-basing. Rejects anything that doesn't parse to a finite number ≥ 1 with a clear error: "row numbers are 1-indexed; got 'startRow=0' in 'A0'. Use A1 (not A0), 1:5 (not 0:5)."

5 new tests cover the malformed-row shapes: `A0`, `A0:B5`, `B5:A0`, `0:1`, `1:0`. 184 tests pass.

**Comment by esokullu** [comment]
@codex pls check again

File: src/chrome/src/agent/sheets-tools.js [hunk 4]
```diff
@@ -0,0 +1,487 @@
...
+  if (typeof range !== 'string' || !range.trim()) {
+    throw new Error(`parseA1: range must be a non-empty string, got ${JSON.stringify(range)}`);
+  }
+  const trimmed = range.trim();
+
+  // Optional Sheet!Range prefix.
+  let sheet = null;
+  let body = trimmed;
+  const bangIdx = trimmed.indexOf('!');
+  if (bangIdx !== -1) {
```
[the remark this answers is not in the document]
Comment (esokullu) [reply src/chrome/src/agent/sheets-tools.js]: Fixed in b43b38d.

Replaced `indexOf('!')` with a quote-aware `_splitSheetRange` helper. If the input starts with `'`, it walks char-by-char to find the matching close quote (treating `''` as an escaped quote per Google Sheets / Excel grammar), then expects `!` immediately after the close. Otherwise it splits at the first `!` as before.

`'Sales!Q1'!A1` and `'It''s!Special'!C5` now both parse correctly. Malformed inputs (`'unterminated!A1`, `'foo'X!A1`, `''!A1`) throw with descriptive errors instead of falling through to "not A1 notation".

5 new tests added; 189 passing.

[the remark this answers is not in the document]
Comment (esokullu) [reply src/chrome/src/agent/sheets-tools.js]: @codex check again pls

File: src/chrome/src/agent/sheets-tools.js [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+
+  if (sCol == null && sRow == null) {
+    throw new Error(`parseA1: range must specify at least one column or row: ${JSON.stringify(range)}`);
+  }
+
+  // Whole column ("A:C" or "A:A") — start/end specify columns but no rows.
+  if (hasEnd && !sRow && !eRow && sCol && eCol) {
+    return {
+      sheet, row: 0, col: startCol,
+      rowCount: -1, colCount: endCol - startCol + 1,
```
[the remark this answers is not in the document]
Comment (esokullu) [reply src/chrome/src/agent/sheets-tools.js]: Caught me — fixed in c20ae66.

Both branches now do the same `Math.min` / `Math.max` normalization the rectangular-range branch already had:
- `C:A` → `colCount: 3` (was `-1`)
- `5:1` → `rowCount: 5` (was `-3`)

2 new tests pinned to the reversed shapes. 191 passing.

[the remark this answers is not in the document]
Comment (esokullu) [reply src/chrome/src/agent/sheets-tools.js]: @codex
check again pls

File: src/chrome/src/agent/sheets-tools.js [hunk 6]
```diff
@@ -0,0 +1,547 @@
...
+// "Sheet2!A1:C3" → { sheet: "Sheet2", row: 0, col: 0, rowCount: 3, colCount: 3 }
+//
+// rowCount/colCount = -1 means "whole column/row" — the backend translates
+// this to the appropriate Ctrl+Shift+End / Ctrl+Shift+arrow gesture.
+//
+// Throws on malformed input. The error message includes the offending
+// input so the caller can surface it to the model verbatim.
+// ─────────────────────────────────────────────────────────────────────────
+
+const A1_PATTERN = /^([A-Z]+)?(\d+)?(?::([A-Z]+)?(\d+)?)?$/;
```
[the remark this answers is not in the document]
Comment (esokullu) [reply src/chrome/src/agent/sheets-tools.js]: @codex please check 0b82ba0 fix: accept absolute A1 references
also pls check all before i merge,
